lambda_builders/lib/base.py

```diff
@@ -0,0 +1,145 @@
+"""
```

Comment:
I'm not sold on the directory structure we have so far. I like my directory structure to be conducive to new contributors finding things easily and making it clear which component is where, and making it easy to extend current functionality. We have to take special care here as well since we have a fold in our code where we need to know about language-specific things, and on the other side we do not. At the high level we have two main abstractions.`Builder` and `Action`. Builder's do not need to have any specific knowledge about the language they are building for, they just encode what they are building in their classname and their capabilities object, as well as containing a list of actions. 

The actions themselves do not need to know anything specific to their language as some will be general purpose like `CopySourceAction`. Others like `PythonPipResolveAction` obviously are specific to python, but don't acutaly have to have any logic in them specific to python, the actions execute methods are essentially just going to proxy out out to the lower level classes that implement language-specific procedures. I think the ladder classes are the only ones that really need to be "below the fold" of language-specific. So what do you think about the project layouts below:

```
aws_lambda_builders
├── __init__.py
├── __main__.py <- CLI hooks
├── action.py   <- `BaseAction`, as well as all of  its subclasses are in here.
├── build.py    <- The main builder is started from here. This is what selects a language-specific builder and runs it.
├── builders.py <- Language specific builders live in here. They can all be registered by doing from builders import *
├── exceptions.py
├── language_support
│   ├── __init__.py
│   ├── dotnet
│   │   └── __init__.py
│   ├── javascript
│   │   └── __init__.py
│   └── python
│       └── __init__.py <- Here is where all the python specific things happen. Even though the `PythonPipResolveAction` lives in the `actions.py` above, it imports pieces from here and runs them. 
└── registry.py 
4 directories, 10 files
```

Downside of this approach is that all the related actions/builders/low level implementation details are not sorted by language, they are instead sorted by their place in the class hierarchy. So adding a new builder would mean doing three things:
* Add a `CoolNewThingBuilder` to the `builders.py` file.
* Add any new `Action` subclasses needed by the new builder to the `actions.py` file.
* Any low level logic needed by the new `Action`s to perform the build of a `CoolNewThing` would go in `language_support/cool_new_thing/__init__.py` (or sibling files) and would be imported by the new actions.


Or what about this
```
aws_lambda_builders
├── __init__.py
├── __main__.py
├── action.py   <- This now just has the BaseAction class and any common language-agnostic Actions like CopySourceAction
├── build.py    <- This is more or less the same as above.
├── builders  <- Now instead of having all the builders/actions in one place, they are sorted by their language. You now need to use your technique to iterate these submodules and import the builders so that they get included in your registry.
│   ├── __init__.py
│   ├── dotnet
│   │   ├── __init__.py
│   │   ├── actions.py
│   │   └── builders.py
│   ├── javascript
│   │   ├── __init__.py
│   │   ├── npm_packager.js <- custom stuff would live alongside the python in these modules or in more subfolders.
│   │   ├── actions.py
│   │   └── builders.py
│   └── python
│       ├── __init__.py
│       ├── packager.py <- low level packager code that is called by the PythonPipResolveAction to do its job.
│       ├── actions.py  <- This now contains python specific actions like PythonPipResolveAction
│       └── builders.py <- The python specific builders would be defined here
├── exceptions.py
└── registry.py

4 directories, 16 files
```

Benefit here is that our high level build/action system is pulled out, and each language-specific piece acts almost like a plugin, its its own self contained directory. Someone could develop their own "package" with the structure
```
ruby
├── __init__.py
├── packager.rb
├── utils.rb
├── actions.py
└── builders.py
```
And essentially drop into the builders package (or maybe we can have a notion of a `BUILDERS_PATH` that is searched for these things and the default entry is this vended builders dir.) to get it to work. This seems the friendliest to me.

Thoughts?

Reply:
This was implemented in #8